src/Dependencies/Collections/ImmutableSegmentedDictionary`2.cs

```diff
@@ -0,0 +1,291 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.CodeAnalysis.Shared.Collections
+{
+    internal partial class ImmutableSegmentedDictionary<TKey, TValue> : IImmutableDictionary<TKey, TValue>, IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>, IDictionary
```

Comment:
This should come with a bunch of documentation: when to use this, what the perf characteristics are, etc.

Reply:
Same for the rest of the dictionary and builder types in this PR. I'm going to hold off on reviewing until this is added, because I'm not sure what I'm looking to review for.

---
In reply to: [541329203](https://github.com/dotnet/roslyn/pull/49536#discussion_r541329203) [](ancestors = 541329203)

Reply:
The performance characteristics are:

* O(1) reads
* O(n) mutations using the immutable single-mutation APIs
* O(n) for the first mutation and O(1) for subsequent mutations when using the builder APIs

This type is _almost_ identical to the one in #47641 except for two differences:

1. The wrapper is a `class` instead of a `struct`
2. This type will not use the Large Object Heap for storage